phylokit/maximum_likelihood/felsenstein.py

```diff
@@ -0,0 +1,275 @@
+        GENOTYPE_ARRAY = np.array(
+            [b"A", b"C", b"G", b"T"],
+            dtype="S",
+        )
+    else:
+        GENOTYPE_ARRAY = np.array(GENOTYPE_ARRAY, dtype="S")
+
+    if pi is None:
+        pi = np.full((4, 4), 0.25, dtype=np.float64)
+
+    likelihoods = _naive_likelihood_felsenstein(
+        ds.node_left_child.data,
+        ds.node_right_sib.data,
+        ds.call_genotype.data,
+        ds.nodes.shape[0],
+        ds.traversal_postorder.data,
+        ds.sample_node.data,
+        util.base_mapping(ds.variant_allele.data, GENOTYPE_ARRAY),
+        ds.node_branch_length.data,
+        rate,
+        pi,
+    )
+
+    ret = np.prod(likelihoods)
+
+    return np.log(ret)
+
+
+@jit.numba_njit()
+def _calculate_likelihood(
+    child,
+    node_branch_length,
+    rate,
+    pi,
+    likelihood,
+):
+    probs = np.zeros(4, dtype=np.float64)
+    for start_state in range(4):
+        for end_state in range(4):
+            probs[start_state] += likelihood[
+                child, end_state
+            ] * _transition_probability(
+                start_state,
+                end_state,
+                node_branch_length[child],
+                rate,
+                pi,
+            )
+    return probs
+
+
+@jit.numba_njit(parallel=True)
+def _likelihood_felsenstein(
+    node_left_child,
+    node_right_sib,
+    call_genotype,
+    num_nodes,
+    traversal_postorder,
+    sample_nodes,
+    variant_allele,
+    node_branch_length,
+    rate,
+    pi,
+):
+    ret = np.zeros(call_genotype.shape[0], dtype=np.float64)
+
+    sample_mask = np.zeros(num_nodes, dtype=np.bool_)
+    sample_mask[sample_nodes] = True
+
+    for i in prange(call_genotype.shape[0]):
+        likelihood = np.zeros((num_nodes, 4), dtype=np.float64)
+        for j, sample_node in enumerate(sample_nodes):
+            if call_genotype[i, j] == -1:
+                likelihood[sample_node] = 0.25
+            else:
+                likelihood[sample_node, variant_allele[i, call_genotype[i, j, 0]]] = 1.0
+        for node in traversal_postorder:
+            if sample_mask[node]:
+                continue
+            else:
+                u = node_left_child[node]
+                probs = _calculate_likelihood(
+                    u,
+                    node_branch_length,
+                    rate,
+                    pi,
+                    likelihood,
+                )
+                v = node_right_sib[u]
+                while v != -1:
+                    probs *= _calculate_likelihood(
+                        v,
+                        node_branch_length,
+                        rate,
+                        pi,
+                        likelihood,
+                    )
+                    v = node_right_sib[v]
+
+                likelihood[node] = probs
+
+        ret[i] = np.sum(likelihood[traversal_postorder[-1]] * 0.25)
+
+    return np.prod(ret)
+
+
+def likelihood_felsenstein(
+    ds,
+    rate,
+    GENOTYPE_ARRAY=None,
```

Comment:
I think we can leave this mapping option out for now, and assume that the states are ACGT